### PR TITLE
E2E Test Scripts in Node

### LINKED ISSUE #120: Enhance E2E Test Scripts
Follow on story to https://github.com/orgs/microsoft/projects/542/views/1.

Enhance E2E testing scripts, edge cases and more thorough check on returned summary object for reporting.

## Acceptance Criteria 
- [ ] Enhanced scripts added to repo. 
- [ ] Check that a particular key is IN_CONSENSUS, NOT_ENOUGH_DATA, LACK_OF_CONSENSUS
- [x] Join the E2E and the Demo scripts together and have the ability to run the Demo in a headless mode so it can be part of CI - #144

### PR DESCRIPTION
**Pull request by macromania**
Fixes #120 

- Implemented a new Typescript NodeJS application as a canary client testing the API end-to-end
- Implemented checks for endpoint responses (200 OK)
- Implemented checks for data report changes (IN_CONSENSUS) etc.

Logs from `make e2e-test` will be like below:

```bash
root@881007c62510:/workspaces/ccf-app-samples/data-reconciliation-app# make e2e-test
npm WARN using --force Recommended protections disabled.
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: eslint-plugin-import@1.11.0
npm WARN Found: eslint@8.31.0
npm WARN node_modules/eslint
npm WARN   dev eslint@"^8.28.0" from the root project
npm WARN   16 more (@typescript-eslint/eslint-plugin, ...)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer eslint@"2.x - 3.x" from eslint-plugin-import@1.11.0
npm WARN node_modules/eslint-plugin-import
npm WARN   dev eslint-plugin-import@"^1.11.0" from the root project
npm WARN 
npm WARN Conflicting peer dependency: eslint@3.19.0
npm WARN node_modules/eslint
npm WARN   peer eslint@"2.x - 3.x" from eslint-plugin-import@1.11.0
npm WARN   node_modules/eslint-plugin-import
npm WARN     dev eslint-plugin-import@"^1.11.0" from the root project

up to date, audited 588 packages in 6s

108 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

> demo-app@1.0.0 start
> ts-node ./src/index.ts




🏁 Starting e2e Tests...

===============================




🤖 Starting Host...

===============================

📀 [Running] - Build App

> build
> del-cli -f dist/ && rollup --config && cp src/endpoints/app.json dist/ && node build_bundle.js dist/


Writing bundle containing 20 modules to dist/bundle.json


📀 [Running] - Start Host

Setting up Python environment...

Python environment successfully setup

[18:37:51.618] Virtual mode enabled

[18:37:51.621] Starting 1 CCF node...

[18:38:13.848] Started CCF network with the following nodes:

[18:38:13.850]   Node [0] = https://127.0.0.1:8000
[18:38:13.851] You can now issue business transactions to the libjs_generic application
[18:38:13.851] Loaded JS application: ./dist/
[18:38:13.852] Keys and certificates have been copied to the common folder: /workspaces/ccf-app-samples/data-reconciliation-app/workspace/sandbox_common
[18:38:13.852] See https://microsoft.github.io/CCF/main/use_apps/issue_commands.html for more information
[18:38:13.853] Press Ctrl+C to shutdown the network

✅ [Done] - Testing Environment Ready!



🔬 [TEST]: Ingestion & Reporting...

===============================

📝 Member 0 Ingesting Data...
✅ [PASS] [200 : OK] - Member 0 data has ingested successfully

📝 Member 1 Ingesting Data...
✅ [PASS] [200 : OK] - Member 1 data has ingested successfully

📝 Member 2 Ingesting Data...
✅ [PASS] [200 : OK] - Member 2 data has ingested successfully

📝 Member 0 Reporting Data...
✅ [PASS] [200 : OK] - Member 0

┌─────────┬─────────────────────┬───────────────────┬────────────────────────┬──────────────────────┬────────────────────────┬───────────┐
│ (index) │    group_status     │ majority_minority │ count_of_unique_values │ members_in_agreement │          lei           │   nace    │
├─────────┼─────────────────────┼───────────────────┼────────────────────────┼──────────────────────┼────────────────────────┼───────────┤
│    0    │  'NOT_ENOUGH_DATA'  │    'Minority'     │           2            │          1           │ '9845002B6B074505A715' │ 'C.10.12' │
│    1    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          2           │ '984500BA57A56NBD3A24' │ 'G.46.77' │
│    2    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500E1B2CA1D4EKG67' │  'A01.1'  │
│    3    │ 'LACK_OF_CONSENSUS' │    'Minority'     │           2            │          1           │ '9845001D460PEJE54159' │ 'K.64.12' │
│    4    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '254900Z8QM2AR51H5I26' │ 'C.10.1'  │
│    5    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500F5BD5BE5767C51' │ 'C.18.13' │
│    6    │ 'LACK_OF_CONSENSUS' │    'Minority'     │           2            │          1           │ '9845005E5AEEB78CE366' │ 'G.46.20' │
│    7    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          2           │ '9845009C73F4BD8X7827' │ 'R.90.01' │
│    8    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500815D6139D53E23' │ 'M.69.1'  │
│    9    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '984500UF3DE41EFA7F02' │ 'B.08.9'  │
└─────────┴─────────────────────┴───────────────────┴────────────────────────┴──────────────────────┴────────────────────────┴───────────┘
📝 Member 1 Reporting Data...
✅ [PASS] [200 : OK] - Member 1

┌─────────┬─────────────────────┬───────────────────┬────────────────────────┬──────────────────────┬────────────────────────┬───────────┐
│ (index) │    group_status     │ majority_minority │ count_of_unique_values │ members_in_agreement │          lei           │   nace    │
├─────────┼─────────────────────┼───────────────────┼────────────────────────┼──────────────────────┼────────────────────────┼───────────┤
│    0    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500E1B2CA1D4EKG67' │  'A01.1'  │
│    1    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '9845001D460PEJE54159' │ 'K.65.12' │
│    2    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '254900Z8QM2AR51H5I26' │ 'C.10.1'  │
│    3    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500F5BD5BE5767C51' │ 'C.18.13' │
│    4    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '9845005E5AEEB78CE366' │ 'G.46.19' │
│    5    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500815D6139D53E23' │ 'M.69.1'  │
│    6    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '984500UF3DE41EFA7F02' │ 'B.08.9'  │
└─────────┴─────────────────────┴───────────────────┴────────────────────────┴──────────────────────┴────────────────────────┴───────────┘
📝 Member 2 Reporting Data...
✅ [PASS] [200 : OK] - Member 2

┌─────────┬─────────────────────┬───────────────────┬────────────────────────┬──────────────────────┬────────────────────────┬───────────┐
│ (index) │    group_status     │ majority_minority │ count_of_unique_values │ members_in_agreement │          lei           │   nace    │
├─────────┼─────────────────────┼───────────────────┼────────────────────────┼──────────────────────┼────────────────────────┼───────────┤
│    0    │  'NOT_ENOUGH_DATA'  │    'Minority'     │           2            │          1           │ '9845002B6B074505A715' │ 'C.10.13' │
│    1    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          2           │ '984500BA57A56NBD3A24' │ 'G.46.77' │
│    2    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500E1B2CA1D4EKG67' │  'A01.1'  │
│    3    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '9845001D460PEJE54159' │ 'K.65.12' │
│    4    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '254900Z8QM2AR51H5I26' │ 'C.10.1'  │
│    5    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500F5BD5BE5767C51' │ 'C.18.13' │
│    6    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '9845005E5AEEB78CE366' │ 'G.46.19' │
│    7    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          2           │ '9845009C73F4BD8X7827' │ 'R.90.01' │
│    8    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500815D6139D53E23' │ 'M.69.1'  │
│    9    │ 'LACK_OF_CONSENSUS' │    'Minority'     │           2            │          1           │ '984500UF3DE41EFA7F02' │ 'B.08.10' │
└─────────┴─────────────────────┴───────────────────┴────────────────────────┴──────────────────────┴────────────────────────┴───────────┘



🔬 [TEST]: Report Changes...

===============================

📝 Member 0 Ingesting Data...
✅ [PASS] [200 : OK] - Member 0 data has ingested successfully

📝 Member 0 Reporting Data...
✅ [PASS] [200 : OK] - Member 0

┌─────────┬─────────────────────┬───────────────────┬────────────────────────┬──────────────────────┬────────────────────────┬───────────┐
│ (index) │    group_status     │ majority_minority │ count_of_unique_values │ members_in_agreement │          lei           │   nace    │
├─────────┼─────────────────────┼───────────────────┼────────────────────────┼──────────────────────┼────────────────────────┼───────────┤
│    0    │  'NOT_ENOUGH_DATA'  │    'Minority'     │           2            │          1           │ '9845002B6B074505A715' │ 'C.10.12' │
│    1    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          2           │ '984500BA57A56NBD3A24' │ 'G.46.77' │
│    2    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500E1B2CA1D4EKG67' │  'A01.1'  │
│    3    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '9845001D460PEJE54159' │ 'K.65.12' │
│    4    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          1           │ '63670082GQTOJV3WDH36' │ 'S.95.21' │
│    5    │ 'LACK_OF_CONSENSUS' │    'Minority'     │           2            │          1           │ '254900Z8QM2AR51H5I26' │ 'C.10.2'  │
│    6    │  'NOT_ENOUGH_DATA'  │    'Majority'     │           1            │          1           │ '391200MF7R7VPUZNDM05' │ 'S.94.20' │
│    7    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500F5BD5BE5767C51' │ 'C.18.13' │
│    8    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '9845005E5AEEB78CE366' │ 'G.46.19' │
│    9    │  'NOT_ENOUGH_DATA'  │    'Minority'     │           2            │          1           │ '9845009C73F4BD8X7827' │ 'R.89.01' │
│   10    │   'IN_CONSENSUS'    │    'Majority'     │           1            │          3           │ '984500815D6139D53E23' │ 'M.69.1'  │
│   11    │ 'LACK_OF_CONSENSUS' │    'Majority'     │           2            │          2           │ '984500UF3DE41EFA7F02' │ 'B.08.9'  │
└─────────┴─────────────────────┴───────────────────┴────────────────────────┴──────────────────────┴────────────────────────┴───────────┘
✅ [PASS] - 12 items in the report
✅ [PASS] - 9845001D460PEJE54159 IN_CONSENSUS group status
✅ [PASS] - 9845001D460PEJE54159 Majority majority/minority
✅ [PASS] - 9845001D460PEJE54159 3 members in agreement

📀 [Running] - Stop Host




🎉 All Tests Passed...

===============================
```

### CODE REVIEW

File: data-reconciliation-app/Makefile [hunk 1]
```diff
@@ -48,7 +48,7 @@ build-enclave: build ## 📦 Build Enclave container image from Dockerfile
 	@../scripts/build_image.sh enclave
 
 e2e-test: ## 📦 Run end to end tests
-	@echo -e "\e[34m Not Implemented \e[0m" || true
+	@cd ./test/e2e-test/ && npm install --force && npm run start
```
Comment (macromania) [inline data-reconciliation-app/Makefile:51]: To make sure the e2e testing app can be run independently, I have added `npm install --force`.

File: data-reconciliation-app/test/e2e-test/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/.eslintrc.json:1]: Added this to make the e2e testing app independent from main app.

File: data-reconciliation-app/test/e2e-test/.prettierrc [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/.prettierrc:1]: Added this to make the e2e testing app independent from main app.

File: data-reconciliation-app/test/e2e-test/package.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/package.json:1]: Added this to make the e2e testing app independent from main app.

File: data-reconciliation-app/test/e2e-test/src/api.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { DemoMemberProps, DemoProps } from './index';
+import axios from 'axios';
+import { SummaryRecordProps } from '../../../src/models/summary-record';
+
+export interface ReportItem {
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/src/api.ts:5]: I realised we are returning a custom response from `/report` and it is not covered by `models`. So, this interface is used for typecasting that response.

File: data-reconciliation-app/test/e2e-test/src/api.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export default class Api {
+    public static async ingest(props: DemoProps, member: DemoMemberProps) {
+        console.log(`📝 ${member.name} Ingesting Data...`);
+
+        const result = await axios.post(props.ingestUrl, member.data, { httpsAgent: member.httpsAgent });
+
+        if(result.status !== 200) {
+            throw new Error(`🛑 [TEST FAILURE]: Unexpected status code: ${result.status}`);
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/src/api.ts:21]: Throwing an error will fail the `make` command and I think this will be reflected to CI/CD actions too.

File: data-reconciliation-app/test/e2e-test/src/host.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                    }
+
+                    resolve();
+                   });
+
+               
+                childProcess.stdout?.on('data', (data) => {
+                    console.log(data);
+
+                    if (data.includes('Press Ctrl+C')) {
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/src/host.ts:49]: Currently there is no other straightforward way of understanding if host is ready.
Alternatively, I could have created a `setTimeout` based or `while(true)` based polling. However, this looked much more practical and quicker to implement.

If host changes their sandbox logs, this code would require a fix.

File: data-reconciliation-app/test/e2e-test/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        const reportItems = await Api.report(this.demoProps, member0);
+
+        if (reportItems.length !== 12) {
+            throw new Error(`🛑 [TEST FAILURE]: Unexpected number of items in the report: ${reportItems.length}`);
+        }
+        else {
+            console.log(`✅ [PASS] - ${reportItems.length} items in the report`);
+        }
+
+        const reportItem = reportItems[3];
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/src/index.ts:80]: More cases can be added here.

I'm naively testing a change in the report with the second ingestion.

File: data-reconciliation-app/test/e2e-test/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+        if(reportItem.group_status !== 'IN_CONSENSUS') {
+            throw new Error(`🛑 [TEST FAILURE]: Unexpected group status ${reportItem.group_status}. Expected: IN_CONSENSUS`);
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/src/index.ts:83]: To avoid making this canary app complicated, I didn't use a testing framework and instead just simply throwing an error if my assertion fails. I think this will be reflected to CI/CD actions too.

File: data-reconciliation-app/test/e2e-test/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (macromania) [inline data-reconciliation-app/test/e2e-test/tsconfig.json:1]: Added this to make the e2e testing app independent from main app.

**Comment by ross-p-smith** [comment]
Maybe I'm missing a pre-step, but if I clone and run `make e2e-test` I get the following error

```
/workspaces/ccf-app-samples/data-reconciliation-app/test/e2e-test/node_modules/ts-node/src/index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/api.ts:2:19 - error TS2307: Cannot find module 'axios' or its corresponding type declarations.
```

Do you think you may need to add an entry here: - `/workspaces/ccf-app-samples/scripts/initialise_devContainer.sh`

File: data-reconciliation-app/Makefile (hunk 1 above)
Comment (ross-p-smith) [reply data-reconciliation-app/Makefile:51]: There is a step that is run when the devcontainer is created. Maybe you could save some time here `/workspaces/ccf-app-samples/scripts/initialise_devContainer.sh`

File: data-reconciliation-app/test/e2e-test/src/api.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    nace: string;
+}
+
+export default class Api {
+    public static async ingest(props: DemoProps, member: DemoMemberProps) {
+        console.log(`📝 ${member.name} Ingesting Data...`);
+
+        const result = await axios.post(props.ingestUrl, member.data, { httpsAgent: member.httpsAgent });
+
+        if(result.status !== 200) {
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/e2e-test/src/api.ts:20]:
```suggestion
        if (result.status !== 200) {
```

File: data-reconciliation-app/test/e2e-test/src/api.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+        console.log(`✅ [PASS] [${result.status} : ${result.statusText}] - ${member.name} ${result.data.content}\n`);
+    }
+
+    public static async report(props: DemoProps, member: DemoMemberProps): Promise<Array<ReportItem>> {
+        console.log(`📝 ${member.name} Reporting Data...`);
+
+        const result = await axios.get(props.reportUrl, { httpsAgent: member.httpsAgent });
+
+        if(result.status !== 200) {
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/e2e-test/src/api.ts:32]:
```suggestion
        if (result.status !== 200) {
```

Linter throwing these up!

File: data-reconciliation-app/test/e2e-test/src/data.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
+export const member0DataPart1 = [
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/e2e-test/src/data.ts:1]: Would these be better as flat files so if Brent gives us a data set we could easily test them?

File: data-reconciliation-app/test/e2e-test/src/host.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
+import { exec, execSync } from 'child_process';
+
+
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/e2e-test/src/host.ts:3]:
```suggestion
```

File: data-reconciliation-app/test/e2e-test/src/host.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    public static stop() {
+        console.log(`\n📀 [Running] - Stop Host\n`);
+
+        /**
+         * Sandbox script spins up the host in a child process using python
+         * We query the python processes and kill it
+         * This makes sure the host is stopped and the sandbox is cleaned up
+         * This also makes the e2e test script idempotent
+         */
+        execSync('kill -9 $(pgrep python)');
```
Comment (ross-p-smith) [inline data-reconciliation-app/test/e2e-test/src/host.ts:28]: Is there a nicer way to do this than killing python? Will you not have the shell script still running in a broken state?

I ask as I did this originally in the bash script and found side-effects of killing python

**Comment by macromania** [comment]
> Maybe I'm missing a pre-step, but if I clone and run `make e2e-test` I get the following error
> 
> ```
> /workspaces/ccf-app-samples/data-reconciliation-app/test/e2e-test/node_modules/ts-node/src/index.ts:859
>     return new TSError(diagnosticText, diagnosticCodes, diagnostics);
>            ^
> TSError: ⨯ Unable to compile TypeScript:
> src/api.ts:2:19 - error TS2307: Cannot find module 'axios' or its corresponding type declarations.
> ```
> 
> Do you think you may need to add an entry here: - `/workspaces/ccf-app-samples/scripts/initialise_devContainer.sh`

@ross-p-smith That was caused by a missing entry in `package.json`. It should be fixed now :)

File: data-reconciliation-app/test/e2e-test/src/host.ts (hunk 7 above)
Comment (julioalex-rezende) [reply data-reconciliation-app/test/e2e-test/src/host.ts:48]: IMHO It's worth to have a comment in the code explaining this.

File: .gitignore [hunk 16]
```diff
@@ -10,3 +10,4 @@ package-lock.json
 **/dist
 **/coverage
 data-reconciliation-app/demo-app/Reports/
+data-reconciliation-app/test/e2e-test/.DS_Store
```
Comment (ross-p-smith) [inline .gitignore:13]:
```suggestion
**/.DS_Store
```

File: data-reconciliation-app/test/e2e-test/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,132 @@
+import Api from './api';
+import https from 'https';
+import fs from 'fs';
+import { member0DataPart1, member0DataPart2, member1Data, member2Data } from './data';
+import Host from './host';
+
+const serverUrl = 'https://127.0.0.1:8000';
+const certificateStorePath = './workspace/sandbox_common';
```
Comment (julioalex-rezende) [inline data-reconciliation-app/test/e2e-test/src/index.ts:8]: If we make this generic to get server and certs_folder based on running mode, we can make this run in all of them.
is it our intention to run in all modes or just sandbox?

Comment (macromania) [reply data-reconciliation-app/test/e2e-test/src/index.ts:8]: In this iteration, that wasn't the intention. For now, I have just provided e2e scripts on a sandbox environment so we can run in CI/CD.

File: data-reconciliation-app/test/e2e-test/src/host.ts (hunk 15 above)
Comment (macromania) [reply data-reconciliation-app/test/e2e-test/src/host.ts:27]: As we discussed in the standup, this can be looked at it in another PR. I have created #142  for that.

File: data-reconciliation-app/test/e2e-test/src/data.ts (hunk 13 above)
Comment (macromania) [reply data-reconciliation-app/test/e2e-test/src/data.ts:1]: I think implementing #131 will help solving this point. I have created a follow up issue: #143

File: data-reconciliation-app/test/e2e-test/src/host.ts (hunk 7 above)
Comment (macromania) [reply data-reconciliation-app/test/e2e-test/src/host.ts:52]: I have updated the code with a comment. #142 will tackle this in another iteration

File: data-reconciliation-app/Makefile (hunk 1 above)
Comment (macromania) [reply data-reconciliation-app/Makefile]: Updated.

File: data-reconciliation-app/test/e2e-test/src/index.ts (hunk 17 above)
Comment (ross-p-smith) [reply data-reconciliation-app/test/e2e-test/src/index.ts:8]: Just IMO - I think this PR needs to fully implement all 3 hosting models. We have working bash scripts that test the application in virtual/docker/mccf. If we are introducing a new mechanism to test (TypeScript) then this needs to cover those same bases. Otherwise we have made it worse - I'm more than willing to help you if you need - but this PR at the moment is not complete. It's not far off finished though! :-)

**Comment by macromania** [comment]
This has been a good experimental approach to e2e testing. It is agreed with the team that for the consistency of the tests with rest of the project, this PR is closed for now. The code will stay in its branch for future ideas and exploration.
